### ISSUE TITLE
refactoring: no import torch in the example notebooks

### ISSUE THREAD

**Issue by jeff-regier** [YosefLab/scVI#87]
As discussed, let's refactor the scvi package so we don't have to `import torch` or anything from torch in the notebooks. That keeps users from having to interact directly with data loaders, and it should make #79 straightforward. One way to achieve this is to add a new class to scVI that contains as members both one dataset and one model. These two thing together define a posterior distribution (implicitly). `Inference` might be a reasonable name for the class. We could put nearly all the scvi code except what's under `models` and `datasets` into this class. `Inference` could either have a `Trainer` instance as a class member, or else `Inference` could subsume `Trainer`. In the latter case, w'd have multiple types of `Inference` classes, like `SemiSupervisedInference`, each with a `train` method. I'm not really sure which way is better.

**Comment by jeff-regier** [comment]
So a basic scvi notebook session might look more like this:

```
gene_dataset = CortexDataset()
vae = VAE(gene_dataset.nb_genes, n_batch=gene_dataset.n_batches * use_batches, n_labels=gene_dataset.n_labels)
infer = Inference(gene_dataset, vae, use_cuda=True, tt_split=0.9)

stats = infer.train(n_epochs=1000, lr=1e-4)

infer.show_t_sne()
imputation_errors = infer.imputation(rate=0.3)
...
```

**Comment by Edouard360** [comment]
perfect working on it !

**Comment by Edouard360** [comment]
it's still on its way but I have started implementing also the Inference class in branch [trainer](/YosefLab/scVI/tree/trainer)

**Comment by Edouard360** [comment]
So I had a question about #79:

As long as we have `--harmonization` and `--annotation`, we override the `--dataset` and `--model` parameters ? And what about the mapping ? Should I define (hardcode) combinations of `dataset` + `model`  in `run_benchmarks` ?

Finally, I think I would be more leaning towards a `.trainer` attribute for the `Inference` class, defined at `__init__` but with a default value so the user can bypass the instanciation.

**Comment by jeff-regier** [comment]
Yes, hard coded combinations could be fine. I'd ignore --dataset if --harmonization or --annotation is  specified. Trainer attribute sounds good.

**Comment by Edouard360** [comment]
## Current progress

As of now, the current design is the following:

#### Inference.
- The Inference class absorbed the Stats class. It should wrap every PyTorch's training procedures, and optionally keeps track of training statistics (and have an early stopping feature on a given metric). Each subclass (SemiSupervisedInference, UnsupervisedInference) can override: .loss(), .on_epoch_end(), ect… 

#### Metrics.
- It seems to me that median imputation error, batch entropy mixing, differential expression are metrics that should stand on the same level as likelihood and accuracy (only the latter were monitored by the stats). We might want to optimise with respect to any of those, and so the metric mode should be known to the Inference class (“max”, or “min”, to indicate which is better). I created a Metric class, with a `.mode` attribute and a \_\_call\_\_ function.

#### Tasks.
- We might have intermediate tasks like `.differential_expression_stats`, or `.imputation_stats`, or comprehensive ones like `.show_t_sne`.

#### Baselines.
- Tasks and Metrics are computed with respect to a PyTorch data_loader. We might have extra metrics that use the raw data like the random forest or support vector machine (`.svc`, or `.rf`) benchmarks.

#### DataLoaders.
- To separate as much the “data-related” logic from the training algorithms in Inference, a DataLoaders class, inherited in TrainTestDataLoaders, JointSemiSupervisedDataLoaders, AlternateSemiSupervisedDataLoaders. Indeed, the training loop can involve the combination of multiple data_loaders at once (for now : 1 or 2). iterates over the data_loaders, preprocess the tensors (.to_cuda), and yield them to the .loss method. A DataLoaders instance is therefore coupled with the loss method, (eg. `AlternateSemiSupervisedVariationalInference` and has `.loss` method with signature `.loss(tensors_unlabelled, tensors_labelled)`, whereas `VariationalInference` has `.loss` method with signature `.loss(tensors)` (only the train tensors))
- Additionally: 
    - a `.to_monitor` attribute for the default DataLoaders on which to keep track of statistics.
    - a `__get_item__` method for simplicity, eg. data_loaders[’train’], data_loaders[’test’], data_loaders[’labelled’]…

For now, tasks and metrics are dynamically set as method of the inference object, and "inference-wise” data loaders can be referred by their name, making the manipulation convenient:

```
cortex_dataset = CortexDataset()
vae = VAE(cortex_dataset.nb_genes)
infer_cortex_vae = VariationalInference(vae, cortex_dataset, train_size=0.1, use_cuda=use_cuda)
infer_cortex_vae.fit(n_epochs=n_epochs)

infer_cortex_vae.ll(’train')
infer_cortex_vae.ll('test')
infer_cortex_vae.differential_expression('test')
infer_cortex_vae.imputation('test', rate=0.1)
infer_cortex_vae.show_t_sne('test', n_samples=1000)
```

## Better design patterns suggestions ?

However the design pattern here is not ideal. I have though of creating an Interface for the Metrics available, like:

```
class VariationalMetrics(LLMetric, ImputationMetric, BatchEntropyMixingMetric):
    pass
```

to bring as many additional method to the inference object, but then I am not sure how to keep track of their “mode” (“max”, or “min”, to indicate which is better)).

**Comment by Edouard360** [comment]
There are also some PyCharm's generated UMLs

<img width="1186" alt="inference" src="https://user-images.githubusercontent.com/15527397/42541863-0aa3dd0c-8459-11e8-9e6c-671c2be0e8d9.png">
<img width="857" alt="dataloaders" src="https://user-images.githubusercontent.com/15527397/42541868-0d4fb7e2-8459-11e8-88fe-82759537be13.png">
<img width="952" alt="metric" src="https://user-images.githubusercontent.com/15527397/42541872-1015a84c-8459-11e8-8fbd-afc1a2ed33be.png">
<img width="593" alt="task" src="https://user-images.githubusercontent.com/15527397/42541876-12915558-8459-11e8-8b2c-da7f86db0e31.png">

**Comment by jeff-regier** [comment]
That all sounds really good. I'm not entirely sure I understand the problem keeping track of "mode". Can we just always minimize metrics? It's always the negative ELBO (or some model) that we're minimizing---we don't want to over-engineer, e.g. planning for a future where we might be optimizing something else. Not sure whether that gets at what you think may be suboptimal about the design.

Anyway, design will always be a work in progress -- maybe good now to get the code committed and to close the issues? I'm glad you've been thinking carefully about the design :-)

**Comment by jeff-regier** [comment]
Ah, I like that dataloaders hierarchy. That's a good idea.

**Comment by Edouard360** [comment]
Thank you Jeffrey ! I am happy you like some part of it ! 😃 

Yes for VAE, we only minimize LL, and we might only monitor this metric for early stopping. 

There is also `ClassifierInference` to train a standard classifier or a classifier on top of a VAE model, for which the optimization metric is the accuracy. And a last detail: I actually implemented as long with the "early\_stopping\_criterion" a “save\_best\_state” criterion, only for the semi-supervised learning. In this setting, the LL might still be used for convergence, but train_accuracy might be used for saving the "best state” of our model - we had training curves where accuracy was slightly oscillating.

Anyway, I have found a simpler way to provide the `infer ` object instance with the wanted methods that made me save ~100 lines. Before merging, I will just need to re-run the notebooks with the new Inference object.
